File: mobile/lib/authValidation.js
const PASSWORD_MIN_LEN = 8;
const USERNAME_RE = /^[a-zA-Z0-9_]{3,32}$/;
const EMAIL_RE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
const HAS_SPECIAL = /[!@#$%^&*()_+\-=[\]{}|;:,.<>?]/;

export function passwordRuleChecks(password) {
  const p = password || '';
  return [
    { id: 'len', label: `Al menos ${PASSWORD_MIN_LEN} caracteres`, ok: p.length >= PASSWORD_MIN_LEN },
    { id: 'upper', label: 'Mayúscula (A–Z)', ok: /[A-Z]/.test(p) },
    { id: 'lower', label: 'Minúscula (a–z)', ok: /[a-z]/.test(p) },
    { id: 'digit', label: 'Un número', ok: /\d/.test(p) },
    { id: 'special', label: 'Carácter especial', ok: HAS_SPECIAL.test(p) },
  ];
}

export function validateRegisterForm({ fullName, email, username, password, passwordConfirm }) {
  const errors = {};
  const name = (fullName || '').trim().replace(/\s+/g, ' ');
  if (name.length < 3) errors.fullName = 'Nombre: mínimo 3 caracteres.';
  const em = (email || '').trim().toLowerCase();
  if (!em) errors.email = 'Correo obligatorio.';
  else if (!EMAIL_RE.test(em)) errors.email = 'Correo no válido.';
  const user = (username || '').trim().toLowerCase();
  if (!user) errors.username = 'Usuario obligatorio.';
  else if (!USERNAME_RE.test(user)) errors.username = 'Usuario: 3–32 caracteres (letras, números, _).';
  if (!passwordRuleChecks(password).every((r) => r.ok)) {
    errors.password = 'La contraseña no cumple todas las reglas.';
  }
  if ((password || '') !== (passwordConfirm || '')) {
    errors.passwordConfirm = 'Las contraseñas no coinciden.';
  }
  return { ok: Object.keys(errors).length === 0, errors };
}

export function validateLoginForm({ username, password }) {
  const errors = {};
  if (!(username || '').trim()) errors.username = 'Introduce tu usuario.';
  if (!(password || '')) errors.password = 'Introduce tu contraseña.';
  return { ok: Object.keys(errors).length === 0, errors };
}

File: backend/auth_validation.py
"""
Validación de registro: usuario, correo y contraseña segura.
"""
from __future__ import annotations

import re
from typing import List, Tuple

EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
USERNAME_RE = re.compile(r"^[a-zA-Z0-9_]{3,32}$")

PASSWORD_MIN_LEN = 8
PASSWORD_SPECIAL = re.compile(r"[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]")


def normalize_email(email: str) -> str:
    return (email or "").strip().lower()


def normalize_username(username: str) -> str:
    return (username or "").strip().lower()


def normalize_full_name(full_name: str) -> str:
    return " ".join((full_name or "").strip().split())


def password_rule_checks(password: str) -> List[Tuple[str, bool]]:
    p = password or ""
    return [
        (f"Al menos {PASSWORD_MIN_LEN} caracteres", len(p) >= PASSWORD_MIN_LEN),
        ("Una letra mayúscula (A–Z)", bool(re.search(r"[A-Z]", p))),
        ("Una letra minúscula (a–z)", bool(re.search(r"[a-z]", p))),
        ("Un número (0–9)", bool(re.search(r"\d", p))),
        ("Un carácter especial (!@#$…)", bool(PASSWORD_SPECIAL.search(p))),
    ]


def validate_password(password: str) -> Tuple[bool, str, List[str]]:
    checks = password_rule_checks(password)
    failed = [label for label, ok in checks if not ok]
    if not failed:
        return True, "ok", []
    hints = [f"La contraseña debe incluir: {', '.join(failed)}."]
    return False, hints[0], hints


def validate_password_confirm(password: str, confirm: str) -> Tuple[bool, str]:
    if (password or "") != (confirm or ""):
        return False, "La confirmación de contraseña no coincide."
    return True, "ok"


def validate_username(username: str) -> Tuple[bool, str]:
    u = normalize_username(username)
    if not u:
        return False, "El nombre de usuario es obligatorio."
    if not USERNAME_RE.match(u):
        return (
            False,
            "Usuario: 3–32 caracteres, solo letras, números y guion bajo (_).",
        )
    return True, "ok"


def validate_email(email: str) -> Tuple[bool, str]:
    e = normalize_email(email)
    if not e:
        return False, "El correo electrónico es obligatorio."
    if not EMAIL_RE.match(e):
        return False, "Introduce un correo electrónico válido."
    return True, "ok"


def validate_full_name(full_name: str) -> Tuple[bool, str]:
    name = normalize_full_name(full_name)
    if len(name) < 3:
        return False, "El nombre completo debe tener al menos 3 caracteres."
    if len(name) > 120:
        return False, "El nombre completo es demasiado largo (máx. 120)."
    return True, "ok"
